File: DP/src/main/java/Solution53.java
public class Solution53 {
    public int maxSubArray(int[] nums) {
        int n = nums.length;

        int[] opt = new int[n];
        opt[0] = nums[0];

        for (int i = 1; i < n; i++) {
            opt[i] = Math.max(opt[i - 1] + nums[i], nums[i]);
        }

        int ans = nums[0];

        for (int o : opt) {
            ans = Math.max(ans, o);
        }

        return ans;

    }
}

File: DP/src/main/python/solution53.py
from typing import List


class Solution:
    def maxSubArray(self, nums: List[int]) -> int:

        n = len(nums)
        opt = [0 for i in range(n)]

        opt[0] = nums[0]

        for i in range(1, n):
            opt[i] = max(opt[i - 1] + nums[i], nums[i])

        # print(opt)
        return max(opt)
